pkg/detectors/netsuite/netsuite.go

```diff
@@ -0,0 +1,253 @@
+							tokenSecret:    tokenSecret,
+							accountID:      accountID,
+						}
+
+						if !isUniqueKeys(cs) {
+							continue
+						}
+
+						s1 := detectors.Result{
+							DetectorType: detectorspb.DetectorType_Netsuite,
+							Raw:          []byte(consumerKey),
+							RawV2:        []byte(consumerKey + consumerSecret),
+						}
+
+						if verify {
+							client := s.client
+							if client == nil {
+								client = defaultClient
+							}
+
+							isVerified, err := verifyCredentials(ctx,
+								client,
+								cs)
+							s1.Verified = isVerified
+							s1.SetVerificationError(err, consumerKey)
+						}
+						results = append(results, s1)
+					}
+				}
+			}
+		}
+	}
+
+	return results, nil
+}
+
+func (s Scanner) Type() detectorspb.DetectorType {
+	return detectorspb.DetectorType_Netsuite
+}
+
+func verifyCredentials(ctx context.Context, client *http.Client, cs credentialSet) (bool, error) {
+	// for url, filter or replace underscore in accountID if needed and lower case the accountID
+	urlAccountId := strings.ToLower(strings.Replace(cs.accountID, "_", "-", -1))
+
+	baseUrl := "https://" + urlAccountId + ".suitetalk.api.netsuite.com"
+
+	const path = "/services/rest/record/v1/metadata-catalog/check"
+
+	// nonce generate
+	nonce, err := netsuiteNonce(11)
+	if err != nil {
+		return false, err
+	}
+
+	signature := makeSignature(http.MethodGet, baseUrl, path, map[string]string{
+		"consumer_key":     cs.consumerKey,
+		"consumer_secret":  cs.consumerSecret,
+		"token_id":         cs.tokenKey,
+		"token_secret":     cs.tokenSecret,
+		"signature_method": "HMAC-SHA256",
+		"timestamp":        strconv.FormatInt(time.Now().Unix(), 10),
+		"nonce":            nonce,
+		"version":          "1.0",
+		"realm":            cs.accountID,
+	})
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, baseUrl+path, nil)
+
+	if err != nil {
+		return false, err
+	}
+
+	// Set required headers
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", signature)
+
+	// Make the request
+	res, err := client.Do(req)
+	if err != nil {
+		if strings.Contains(err.Error(), "no such host") {
+			return false, nil
+		}
+		return false, err
+	}
+	defer res.Body.Close()
+	switch res.StatusCode {
+	case http.StatusOK:
+		return true, nil
+	case http.StatusUnauthorized:
+		// 401 indicates unverified credentials
+		return false, nil
+	default:
+		return false, fmt.Errorf("unexpected HTTP response status %d", res.StatusCode)
+	}
+}
+
+func generateHMACSHA256(message, secret string) string {
+	// Create a new HMAC by defining the hash type and the key (as byte array)
+	h := hmac.New(sha256.New, []byte(secret))
+
+	// Write the message to it
+	h.Write([]byte(message))
+
+	// Get the resulting HMAC as a byte slice
+	hash := h.Sum(nil)
+
+	// Encode the byte slice to a hexadecimal string
+	return base64.StdEncoding.EncodeToString(hash)
+}
+
+func makeSignature(method, baseUrl, path string, params map[string]string) string {
```

Comment:
question/suggestion: Do you think we should use a strings.Builder here to avoid all these concatenations? It's a little more verbose but it would be more performant. I'm not sure if the tradeoff is worth it?

Something like:
```
func addOAuthParam(builder *strings.Builder, key, value string) {
	builder.WriteString(key)
	builder.WriteString("=\"")
	builder.WriteString(value)
	builder.WriteString("\",")
}

func makeSignature(method, baseUrl, path string, params map[string]string) string {
	var baseStringBuilder strings.Builder

	baseStringBuilder.WriteString(method)
	baseStringBuilder.WriteString("&")
	baseStringBuilder.WriteString(url.QueryEscape(baseUrl + path))
	baseStringBuilder.WriteString("&")
	baseStringBuilder.WriteString(url.QueryEscape(
		"oauth_consumer_key=" + url.QueryEscape(params["consumer_key"]) +
			"&oauth_nonce=" + url.QueryEscape(params["nonce"]) +
			"&oauth_signature_method=" + url.QueryEscape(params["signature_method"]) +
			"&oauth_timestamp=" + url.QueryEscape(params["timestamp"]) +
			"&oauth_token=" + url.QueryEscape(params["token_id"]) +
			"&oauth_version=" + url.QueryEscape(params["version"]),
	))

	baseString := baseStringBuilder.String()
	key := url.QueryEscape(params["consumer_secret"]) + "&" + url.QueryEscape(params["token_secret"])

	signature := generateHMACSHA256(baseString, key)

	var authHeaderBuilder strings.Builder

	authHeaderBuilder.WriteString("OAuth realm=\"")
	authHeaderBuilder.WriteString(params["realm"])
	authHeaderBuilder.WriteString("\",")

	addOAuthParam(&authHeaderBuilder, "oauth_consumer_key", params["consumer_key"])
	addOAuthParam(&authHeaderBuilder, "oauth_token", params["token_id"])
	addOAuthParam(&authHeaderBuilder, "oauth_signature_method", params["signature_method"])
	addOAuthParam(&authHeaderBuilder, "oauth_timestamp", params["timestamp"])
	addOAuthParam(&authHeaderBuilder, "oauth_nonce", params["nonce"])
	addOAuthParam(&authHeaderBuilder, "oauth_version", params["version"])
	addOAuthParam(&authHeaderBuilder, "oauth_signature", url.QueryEscape(signature))

	// Remove the trailing comma
	authHeader := authHeaderBuilder.String()
	authHeader = authHeader[:len(authHeader)-1]

	return authHeader
}
```

Reply:
Using string builder is making code readability difficult. I am also not that sure about the performance.